### PR TITLE
AP-909 Customize number of cores to use for parallelisation in FastSync

### PR DESCRIPTION
**Pull request by Samira-El**
## Problem

FastSync currently uses the number of CPU cores as the size of processes pool to use to sync tables.
This means that this might overwhelm sources that cannot handle multiple processes trying to fetch data at the same time, e.g  databases with low IOPS.

## Proposed changes

Add ability to customize the size of the process pool. 

That is done by adding a source level property `fastsync_parallelism` in the pipelines config to set the size.

## Types of changes

What types of changes does your code introduce to PipelineWise?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] Description above provides context of the change
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Unit tests for changes (not needed for documentation changes)
- [x] CI checks pass with my changes
- [x] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [x] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [x] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions

### CODE REVIEW

File: pipelinewise/fastsync/mongodb_to_postgres.py [hunk 1]
```diff
@@ -110,7 +110,7 @@ def sync_table(table: str, args: Namespace) -> Union[bool, str]:
 def main_impl():
     """Main sync logic"""
     args = utils.parse_args(REQUIRED_CONFIG_KEYS)
-    cpu_cores = utils.get_cpu_cores()
+    cpu_cores = args.tap.get('fastsync_parallelism', utils.get_cpu_cores())
```
Comment (koszti) [inline pipelinewise/fastsync/mongodb_to_postgres.py:113]: would be great to eliminate that so many duplicates in `x_to_y.py` files one day.

File: pipelinewise/cli/schemas/tap.json [hunk 2]
```diff
@@ -22,6 +22,15 @@
         "target_schema"
       ]
     },
+    "db_conn": {
+      "type": "object",
+      "properties": {
+        "fastsync_parallelism": {
+          "type": "integer",
+          "minimum": 1
```
Comment (koszti) [inline pipelinewise/cli/schemas/tap.json:30]: can we have a realistic max value here?

File: install.sh [hunk 3]
```diff
@@ -59,7 +59,7 @@ make_virtualenv() {
     echo "Making Virtual Environment for [$1] in $VENV_DIR"
     python3 -m venv $VENV_DIR/$1
     source $VENV_DIR/$1/bin/activate
-    python3 -m pip install --upgrade pip setuptools
+    python3 -m pip install --upgrade pip setuptools wheel
```
Comment (koszti) [inline install.sh:62]: not necessarily related to the topic in the PR, but ok

File: install.sh [hunk 4]
```diff
@@ -74,8 +74,6 @@ make_virtualenv() {
     fi
 
     echo ""
-    echo "===== Checking dependencies for conflict ..."
-    python3 -m pip check && echo "No conflicts" || exit 1
```
Comment (koszti) [inline install.sh:78]: same as above. Would be better to put it into a separated PR and keep the commit history cleaner, but we can squeeze it into this one as well I think.

File: pipelinewise/fastsync/mongodb_to_postgres.py (hunk 1 above)
Comment (Samira-El) [reply pipelinewise/fastsync/mongodb_to_postgres.py:113]: Indeed, so many files with same code!
Soon Peter, soon!

File: pipelinewise/cli/schemas/tap.json (hunk 2 above)
Comment (Samira-El) [reply pipelinewise/cli/schemas/tap.json:30]: Like what?
`multiprocessing.Pool` does accept 1 as the min size of the pool.

File: install.sh (hunk 3 above)
Comment (Samira-El) [reply install.sh:62]: True lol.
I don't want to make a separate PR for a small change, esp given how long it take for the CI to run.

File: install.sh (hunk 4 above)
Comment (Samira-El) [reply install.sh:78]: should I revert this change

File: pipelinewise/cli/schemas/tap.json (hunk 2 above)
Comment (koszti) [reply pipelinewise/cli/schemas/tap.json:30]: Setting `fastsync_parallelism` to 1000000 likely will cause a death on the host. Do we want to control it by setting a realistic max allowed value in the config, like 100 or we don't want to care and we trust the user and the OS?

File: install.sh (hunk 4 above)
Comment (koszti) [reply install.sh:78]: nah, just a note

File: pipelinewise/cli/schemas/tap.json (hunk 2 above)
Comment (Samira-El) [reply pipelinewise/cli/schemas/tap.json:30]: Wdyt about programmatically falling back to the number of available CPU cores if the given `fastsync_parallelism` is greater than the # cup cores?

Comment (koszti) [reply pipelinewise/cli/schemas/tap.json:30]: sounds good to me

Comment (Samira-El) [reply pipelinewise/cli/schemas/tap.json]: Made few updates:
* set maximum to 1000 ..we could re-consider it if anyone every uses a super powerful node and want to use 1000 cores. :P
* choose min between cpu cores and fastsync_parallelism

**Review by koszti — DISMISSED** [review]

File: pipelinewise/cli/schemas/tap.json (hunk 2 above)
Comment (koszti) [reply pipelinewise/cli/schemas/tap.json]: 🙇

**Comment by Samira-El** [comment]
@koszti Can you bless me with your seal of approval?

**Review by koszti — APPROVED** [review]

**Pull request merged**